### PR TITLE
Node nfc reader

### PR DESCRIPTION
**Pull request by coenhezemans**
nfc reader functional according to design (part of design was unusable and was ignored)

### CODE REVIEW

File: Source/Node/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (NathanThus) [inline Source/Node/.vscode/settings.json:1]: ?

File: Source/Node/include/MFRC522Reader.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#ifndef MFRC522READER_HPP
```
Comment (NathanThus) [inline Source/Node/include/MFRC522Reader.hpp:1]: Do these defines need to be public?

**Review by NathanThus — COMMENTED** [review]
See comments

File: Source/Node/include/MFRC522Reader.hpp (hunk 2 above)
Comment (coenhezemans) [reply Source/Node/include/MFRC522Reader.hpp]: no, removed the public defines and system is still functional

**Comment by TomVer99** [comment]
ERROR: Style Error
Line 67 in /home/runner/work/FindIt/FindIt/Source/Node/src/MFRC522Reader.cpp, Incorrect operator.
return -1;
	Operators must be preceded and followed by a whitespace.


ERROR: Style Error
Line [72](https://github.com/T3-FindIt/FindIt/actions/runs/5001681970/jobs/8960673211?pr=86#step:5:73) in /home/runner/work/FindIt/FindIt/Source/Node/src/MFRC522Reader.cpp, Incorrect operator.
return -1;
	Operators must be preceded and followed by a whitespace.


ERROR: Style Error
Line 84 in /home/runner/work/FindIt/FindIt/Source/Node/src/MFRC522Reader.cpp, Incorrect operator.
return -1;
	Operators must be preceded and followed by a whitespace.


These 3 errors are FALSE POSITIVES. They belong to a bug already reported (#83 ). Other dont seem like false positives.

File: Source/Node/include/MFRC522Reader.hpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#include "IReader.hpp"
+
+#define DEFAULT_SS_PIN 10
+#define DEFAULT_RST_PIN 9
+#define DEFAULT_KEY 0xFF
+#define DEFAULT_TRAILER_BLOCK 7
+
+class MFRC522Reader : public IReader
+{
+    private:
```
Comment (TomVer99) [inline Source/Node/include/MFRC522Reader.hpp:12]: Watch `private:` and `public:` indentation

File: Source/Node/include/MFRC522Reader.hpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    MFRC522::MIFARE_Key key;
+    bool AuthenticateCard();
+    public:
+    MFRC522Reader();
+    MFRC522Reader(int SS_PIN, int RST_PIN);
+    ~MFRC522Reader();
+    bool CheckForCard();
+    // outputstring should be [49]
+    MFRC522::StatusCode status;
+    int ReadCard(char* outputString);
```
Comment (TomVer99) [inline Source/Node/include/MFRC522Reader.hpp:23]: No length variable / paramter?

File: Source/Node/src/MFRC522Reader.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+    else
+    {
+        reader.PICC_IsNewCardPresent();
+        return reader.PICC_ReadCardSerial();
+    }
+}
+
+// outputstring should be [49]
+int MFRC522Reader::ReadCard(char* outputString)
```
Comment (TomVer99) [inline Source/Node/src/MFRC522Reader.cpp:62]: What if outputString is nullptr?

**Review by TomVer99 — CHANGES_REQUESTED** [review]
Talking with @NathanThus about how to properly structure the project, he recently found out a specific structure had to be kept in order for tests to be able to work (I understand that I2C / NFC code cannot be unit tested within the project time span).

File: Source/Node/include/MFRC522Reader.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#include <MFRC522.h>
+
+#include "IReader.hpp"
+
+#define DEFAULT_SS_PIN 10
```
Comment (NathanThus) [inline Source/Node/include/MFRC522Reader.hpp:5]: Does this need to be public?

File: Source/Node/include/MFRC522Reader.hpp (hunk 3 above)
Comment (NathanThus) [reply Source/Node/include/MFRC522Reader.hpp:12]: Ditto

File: Source/Node/include/MFRC522Reader.hpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    private:
+    MFRC522 reader;
+    MFRC522::MIFARE_Key key;
+    bool AuthenticateCard();
+    public:
+    MFRC522Reader();
+    MFRC522Reader(int SS_PIN, int RST_PIN);
+    ~MFRC522Reader();
+    bool CheckForCard();
+    // outputstring should be [49]
```
Comment (NathanThus) [inline Source/Node/include/MFRC522Reader.hpp:21]: Should != is

File: Source/Node/include/MFRC522Reader.hpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    MFRC522::StatusCode status;
+    int ReadCard(char* outputString);
+};
```
Comment (NathanThus) [inline Source/Node/include/MFRC522Reader.hpp:24]: New line here

**Review by NathanThus — CHANGES_REQUESTED** [review]

File: Source/Node/src/MFRC522Reader.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
+#include "MFRC522Reader.hpp"
+#include <SPI.h>
+
+
+#define DEFAULT_SS_PIN 10
+#define DEFAULT_RST_PIN 9
+#define DEFAULT_KEY 0xFF
+#define DEFAULT_TRAILER_BLOCK 7
+
+MFRC522Reader::MFRC522Reader()
```
Comment (TomVer99) [inline Source/Node/src/MFRC522Reader.cpp:10]: Why not call the other constructor? This code is the same except that the SS and RST pins are using the default values. Or maybe only 1 constructor with default parameter values.

File: Source/Node/src/MFRC522Reader.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+{
+    reader = MFRC522(DEFAULT_SS_PIN, DEFAULT_RST_PIN);
+    SPI.begin();
+    reader.PCD_Init();
+    for (byte i = 0; i < 6; i++)
+    {
+        key.keyByte[i] = 0xFF;
```
Comment (TomVer99) [inline Source/Node/src/MFRC522Reader.cpp:17]: = 0xFF, but this value is define above.

File: Source/Node/src/MFRC522Reader.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+
+#define DEFAULT_SS_PIN 10
+#define DEFAULT_RST_PIN 9
+#define DEFAULT_KEY 0xFF
+#define DEFAULT_TRAILER_BLOCK 7
+
+MFRC522Reader::MFRC522Reader()
+{
+    reader = MFRC522(DEFAULT_SS_PIN, DEFAULT_RST_PIN);
```
Comment (TomVer99) [inline Source/Node/src/MFRC522Reader.cpp:12]: What if it does not init correctly? I dont see error handling. Is this not possible?

**Review by TomVer99 — CHANGES_REQUESTED** [review]

File: Source/Node/src/MFRC522Reader.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+MFRC522Reader::MFRC522Reader(int SS_PIN, int RST_PIN)
+{
+    reader = MFRC522(SS_PIN, RST_PIN);
+    SPI.begin();
+    reader.PCD_Init();
+    for (byte i = 0; i < 6; i++)
+    {
+        key.keyByte[i] = 0xFF;
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:28]: More mentions of 0xFF, but it was defined above.

Also, what is that 6 doing there?

File: Source/Node/src/MFRC522Reader.cpp (hunk 11 above)
Comment (coenhezemans) [reply Source/Node/src/MFRC522Reader.cpp:12]: cant use error handling because of void type initializers

File: Source/Node/src/MFRC522Reader.cpp [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    return true;
+}
+
+bool MFRC522Reader::CheckForCard()
+{
+    if (status == MFRC522::STATUS_TIMEOUT)
+    {
+        reader.PCD_Reset();
+        delay(50);
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:52]: Magic numbers!

File: Source/Node/src/MFRC522Reader.cpp [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        reader.PICC_IsNewCardPresent();
+        return reader.PICC_ReadCardSerial();
+    }
+}
+
+// outputstring has to be [49]
+int MFRC522Reader::ReadCard(char* outputString, int arrayLength)
+{
+    if (arrayLength < 49)
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:70]: More magic numbers!

File: Source/Node/src/MFRC522Reader.cpp [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    if (!AuthenticateCard())
+    {
+        return -1;
+    }
+
+    for (int i = 2; i >= 0; i--)
+    {
+        byte readBuffer[18];
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:87]: Even more magic numbers!

File: Source/Node/src/MFRC522Reader.cpp [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        byte size = sizeof(readBuffer);
+
+        byte address = 4 + i;
+        status = (MFRC522::StatusCode)reader.MIFARE_Read(address, readBuffer, &size);
+        if (status != MFRC522::STATUS_OK)
+        {
+            return -1;
+        }
+        for (int j = 0; j < 16; j++)
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:96]: See previous comment(s)

File: Source/Node/src/MFRC522Reader.cpp [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            if (readBuffer[j] == '\n')
+            {
+                outputString[i * 16 + j] = '\0';
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:100]: See previous comment(s)

File: Source/Node/src/main.cpp [hunk 18]
```diff
@@ -1,19 +1,39 @@
...
-    comToHub->SendNewItemToHub(sendstring);
-    digitalWrite(LED_BUILTIN, HIGH);
-    delay(1000);
-    digitalWrite(LED_BUILTIN, LOW);
-    delay(1000);
+    static char currentCard[CARDSTRINGLENGTH];
+    bool status = nfcReader->CheckForCard();
+    if (status)
+    {
+        char card[49];
```
Comment (NathanThus) [inline Source/Node/src/main.cpp:23]: You have a define at line 7 with this value. Why use a magic number here?

File: Source/Node/src/MFRC522Reader.cpp [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+MFRC522Reader::~MFRC522Reader()
+{
+    SPI.end();
+}
+
+bool MFRC522Reader::AuthenticateCard()
+{
+    status = (MFRC522::StatusCode)reader.PCD_Authenticate(MFRC522::PICC_CMD_MF_AUTH_KEY_A, DEFAULT_TRAILER_BLOCK, &key, &(reader.uid));
+    if (status != MFRC522::STATUS_OK)
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:40]: I feel like this can be simplified.

File: Source/Node/src/MFRC522Reader.cpp [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+            if (readBuffer[j] == '\n')
+            {
+                outputString[i * 16 + j] = '\0';
+            }
+            else
+            {
+                outputString[i * 16 + j] = readBuffer[j];
+            }
+        }
+        outputString[48] = '\0';
```
Comment (NathanThus) [inline Source/Node/src/MFRC522Reader.cpp:107]: Even more magical numbers!

**Review by NathanThus — CHANGES_REQUESTED** [review]

**Review by TomVer99 — APPROVED** [review]

**Review by NathanThus — APPROVED** [review]

**Pull request merged**